### ISSUE TITLE
[skrifa] Location/LocationRef type ergonomics

### ISSUE THREAD

**Issue by dfrg** [googlefonts/fontations#866]
The `Location` and `LocationRef` types were intended to serve a similar purpose to other owned/borrowed type pairs in Rust (`String`/`&str`, `Vec<T>`/`&[T]`, etc) but actual usage doesn't actually follow the same patterns leading to confusion and frustration.

The main issue is that `Location` cannot impl `Deref<Target = LocationRef>` due to the requirements of the `Deref` trait. Here's one potential solution along with some proposed renaming for consistency and to make the purpose of these types more clear at a glance:

```rust
/// Coordinate that represents the location of a single axis in variation
/// space, normalized to the range `-1.0..=1.0`.
pub type VarCoord = F2Dot14;

/// Borrowed type for a collection of normalized coordinates, one for each
/// axis in the same order as the `fvar` table, which represents a location
/// in variation space.
///
/// For the owned version of this type, see [`VarLocation`].
pub type VarLocationRef = [VarCoord];

/// Owned type for a collection of normalized coordinates, one for each
/// axis in the same order as the `fvar` table, which represents a location
/// in variation space.
/// 
/// (Other docs for how to construct this, examples, etc)
///
/// For the borrowed version of this type, see [`VarLocationRef`].
#[derive(Default)]
pub struct VarLocation(SmallArray<VarCoord>);

impl Deref for VarLocation {
    type Target = VarLocationRef;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
```

This renames `NormalizedCoord` -> `VarCoord` to make the relation clear and adds a `Var` prefix to the location types.

Notably `VarLocationRef` now just becomes a type alias for `[VarCoord]` which is an unsized type similar to `str` leading to the same usage patterns (you can never use it without a `&`).

All types that accept a location would now take a `&VarLocationRef` parameter which can be passed as follows:

```rust
pub fn uses_variations(var_loc: &VarLocationRef) {}

// with an owned location
let owned_loc = font.axes().location(...);
uses_variations(&owned_loc);

// vello style with coords accumulated in a scene
struct Scene {
    // ...
    all_coords: Vec<VarCoord>,
}
uses_variations(&scene.all_coords[start..end]);

// passing default (or empty) locations
uses_variations(Default::default());
uses_variations(&VarLocation::default());
```

This change can be done incrementally without breaking Skia.

We _could_ make `VarLocationRef` itself a real unsized type to also enable this pattern but that would require unsafe code and IMO a borrowed location is really just a slice of coordinates.

@rsheeter wdyt? @cmyr thoughts on general rustiness of this approach?

**Comment by rsheeter** [comment]
Seems like a lateral pass. "VarCoord" seems like it should be an enum that supports _at least_ user and normalized, having it mean normalized only is a bit weird.

The problem with Location/LocationRef is lack of documentation and non-obvious construction. We added some documentation, for example https://docs.rs/skrifa/0.18.0/skrifa/instance/struct.Location.html now actually explains how to create one, so ... what if we just Do Nothing?

I'm not sure I understand the Vello scene example, how does `all_coords: Vec<VarCoord>` work if you have multiple bits of text with different positions, potentially using fonts that define the same axis differently?

**Comment by cmyr** [comment]
it's an interesting idea, and not something I've ever really seen before (specifically the type alias to the unsized slice type). Would be curious to have someone even weedsier than I on the top of rust API design chime in.

Ultimately I don't really _use_ this API much(/at all) so I don't have a great sense of what the papercuts are. In what cases are you using one vs the other type? Would it make sense to just make it look like a `Cow` or something, where there are owned and borrowed variants?

Otherwise it might be possible to solve this problem with the right function signatures... like you could have all your fns take `location: impl AsRef<[NormalizedCoord]>` (or a custom trait if we wanted to limit the implementors to only our location types).

Those are just two ideas for other possible approaches here, if the current situation is annoying. You could also rip out the current code and go back to always using `&[NormalizedCoord]`, if the simplicity feels like the right tradeoff.

**Comment by rsheeter** [comment]
The tl;dr of the papercuts is that while writing code to do things like make an svg of a glyph at some location in variation space I found myself puzzled on multiple occasions and felt that to a new user this would be even more puzzling.

I have updated the docs in the specific places I got confused, for example to make https://docs.rs/skrifa/latest/skrifa/instance/struct.LocationRef.html and https://docs.rs/skrifa/latest/skrifa/instance/struct.Location.html point the user to AxisCollection::location. Previously if you had some user coords and knew you wanted a Location it wasn't terribly obvious what to do. Perhaps that's sufficient and we can just leave it?

Location/LocationRef didn't (don't?) convert in the manner PathBuf/Path or other comparables do.
